source/common/http/filter/lua/wrappers.cc

```diff
@@ -0,0 +1,90 @@
+#include "common/http/filter/lua/wrappers.h"
+
+namespace Envoy {
+namespace Http {
+namespace Filter {
+namespace Lua {
+
+HeaderMapIterator::HeaderMapIterator(HeaderMapWrapper& parent) : parent_(parent) {
+  entries_.reserve(parent_.headers_.size());
+  parent_.headers_.iterate(
+      [](const HeaderEntry& header, void* context) -> HeaderMap::Iterate {
+        HeaderMapIterator* iterator = static_cast<HeaderMapIterator*>(context);
+        iterator->entries_.push_back(&header);
+        return HeaderMap::Iterate::Continue;
+      },
+      this);
+}
+
+int HeaderMapIterator::luaPairsIterator(lua_State* state) {
+  if (current_ == entries_.size()) {
+    parent_.iterating_ = false;
+    return 0;
+  } else {
+    lua_pushstring(state, entries_[current_]->key().c_str());
+    lua_pushstring(state, entries_[current_]->value().c_str());
+    current_++;
+    return 2;
+  }
+}
+
+int HeaderMapWrapper::luaAdd(lua_State* state) {
+  checkModifiable(state);
+
+  const char* key = luaL_checkstring(state, 2);
+  const char* value = luaL_checkstring(state, 3);
+  headers_.addCopy(LowerCaseString(key), value);
+  return 0;
+}
+
+int HeaderMapWrapper::luaGet(lua_State* state) {
+  const char* key = luaL_checkstring(state, 2);
+  const HeaderEntry* entry = headers_.get(LowerCaseString(key));
+  if (entry != nullptr) {
+    lua_pushstring(state, entry->value().c_str());
+    return 1;
+  } else {
+    return 0;
+  }
+}
+
+int HeaderMapWrapper::luaPairs(lua_State* state) {
+  if (iterating_) {
+    luaL_error(state, "cannot create a second iterator before completing the first");
+  }
+
+  // The way iteration works is we create an iteration wrapper that snaps pointers to all of
+  // the headers. We don't allow modification while an iterator is active. This means that
```

Comment:
Still wrapping my head around the Lua memory model here. Can the underlying objects/strings (in the `HeaderMap`) that have had their pointers snapped be destroyed between when `luaPairs` is called and `luaPairsIterator` happens?

Reply:
Essentially, when `pairs()` is called, we return a lambda. This lambda is called repeatedly by the script until it returns nil. However, a bad script could do something silly like call `pairs()` outside of iteration context. I.e., the following code is valid Lua code:

```
headers_it = headers:pairs()
val1 = headers_it()
val2 = headers_it()
```

It's also possible to break out of a for loop without completing the iteration. In this case, we don't know when the iterator will be destroyed by GC. There are no guarantees. In general I've written the code to be super paranoid to avoid issues from bad scripts, at the expense of being overly strict. This is related to @rshriram comment where it might be also good to have a `clone()` and `merge(headers)` API for headers. I think that's reasonable but I would prefer to wait to get feedback on the initial code. The warnings in the docs are super clear that the API for Lua is not locked yet.

Reply:
I'm thinking specifically about multi-tenant use cases here. E.g. you are a service provider operating a proxy for customer A and B. Should we be aiming to provide guarantees that A can never inject a Lua script that might cause security concerns to B? If so, we need to prevent scripts from ever being able to do bad things.

Reply:
This would require a lot more thinking. In general I think allowing customers to inject Lua code in a multi-tenant environment is probably a non-starter, if for the only reason that a script can spin and we have no real way to break out of it. With that said, in the current model, as long as this is done on a per-listener basis, the Lua jails will be discrete.

Reply:
Well, they will be discrete, but if it's possible to get a dangling `HeaderMap` reference via inappropriate iterator usage, then all bets are off. I'm not really asking that we solve this in this PR, but we probably should clarify in the docs what the MT story is with Envoy. It seems that Envoy is well positioned for MT scenarios functionally for most things other than this filter, modulo QoS and other latency/bandwidth isolation concerns.

Reply:
That part is actually working. It can't be dangling. The way the filter uses the header is via a `DeathRef` which blocks usage outside of appropriate contexts. It will make more sense when you see that PR.